### ISSUE TITLE
Convergence issues when using pytorch's native AMP

### ISSUE THREAD

**Issue by quinor** [pytorch/pytorch#38788]
## 🐛 Bug

The issue has got very weird symptoms.

I've got a pretty complex convolutional network implemented in Pytorch with Lightning (https://github.com/PyTorchLightning/pytorch-lightning). The network trains fine on fp32 (light blue line on the plot below), but when I switch to AMP fp16 with `O2` optimization level (implementation from Lightning), the loss function does something very non-monotonic (dark-red line on the plot). It is important to say that this is training loss, not validation loss.

![image](https://user-images.githubusercontent.com/3879392/82470787-11bca800-9ac6-11ea-9080-ad5a2bd25e34.png)

I could see that the network could just not train properly on AMP, but here the things get much weirder. The network trains correctly in fp16 when I change a specific upscaling in my network from bilinear:

```
nn.Upsample(scale_factor=4*i, mode='bilinear', align_corners=True)
```

to nearest neighbor:

```
nn.Upsample(scale_factor=4*i)
```

That leads me to believe that something very weird may be going on with AMP or fp16 in general. I was trying to force the nn.Upsample to produce a suspicious behavior, but I was unable to do so.

## Expected behavior

The network should converge the same way with and without fp16 enabled.

## Environment

```
Collecting environment information...
PyTorch version: 1.6.0.dev20200516
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
CMake version: version 2.8.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: Tesla V100-PCIE-32GB
Nvidia driver version: 440.64.00
cuDNN version: Probably one of the following:
/apps/gpu/cuda/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/apps/gpu/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/apps/gpu/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/apps/gpu/cuda/cuda-9.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] pytorch-lightning==0.7.6
[pip3] torch==1.6.0.dev20200516
[pip3] torchvision==0.7.0.dev20200516
[conda] Could not collect
```

(cuda 10.2 is used)

## Additional context

@mcarilli

**Comment by ptrblck** [comment]
Hi @quinor,

it seems you are using apex/amp with `opt_level='O2'`, not native amp (via `torch.cuda.amp`).
Could you install the nightly binaries and try Lightning's native amp support, as it should be already available per https://github.com/PyTorchLightning/pytorch-lightning/pull/1561 ?

Also, thanks for narrowing down the potential issue to the upsampling layer.
Could you post some information regarding your setup, i.e. are you using `DistributedDataParallel`, channels-last, etc?
Would you also be able to post the model architecture? If not, could you mention which layers you were using?

**Comment by quinor** [comment]
Hi @ptrblck,

I am pretty convinced I should be using the native amp - as you can see I am running on the nightly binaries and lightning's amp implementation. Or did I not turn it on properly, where can you see that in the logs?

I am not using channels-last as it isn't supported yet. I am not using DDP. The model is pretty complex and I could send it to you privately if you want. As for the layers, I'm mostly using convolutions, dilated convolutions, grouped convolutions, depthwise-separable convolutions, interpolate, batchnorm and relu.

**Comment by ptrblck** [comment]
I'm not seeing any logs, but referred to your information:
> but when I switch to AMP fp16 with O2 optimization level (implementation from Lightning)

If you are using `opt_level='O2'`, then you would use apex/amp, not the PyTorch native implementation, as there are no opt levels.

"The model is pretty complex and I could send it to you privately if you want."
If your model code is not publishable in its current state, I would ask you not to send it to me for now.

We'll take a look at the upsampling layers and try to narrow it down.

**Comment by quinor** [comment]
Right, I believe this option is irrelevant then - I set it up in Lightning and didn't realize it's not passed to the native amp.

As for the model, the code has already been shared with Nvidia for different reasons, so it might be a feasible option if needed.

**Comment by ptrblck** [comment]
@quinor Could you share the nvbugs id, please?
